### ISSUE TITLE
Works with webflux?

### ISSUE THREAD

**Issue by favetelinguis** [aws/aws-xray-sdk-java#54]
Are there any limitations for using aws-xray-recorder-sdk-spring on a webflux project?

**Comment by chrisradek** [comment]
@favetelinguis 
We haven't tested the `aws-xray-recorder-sdk-spring` package with a webflux project, so it will depend on how well webflux plays with spring aop.

Have you tried using the two together, and have you seen any issues?

**Comment by favetelinguis** [comment]
No I have not im planing on setting up X-Ray for a ECS webflux fargate instance and wanted to check if its worth the effort.

**Comment by stevegoossens** [comment]
@chrisradek, presumably anything fundamental in this SDK that implements synchronous/blocking Filter or Servlet features based on `java.servlet.*` (e.g. `AWSXRayServletFilter`) would need an asynchronous/non-blocking equivalent? I think it a non-blocking `AWSXRayServletFilter` would have to implement https://github.com/spring-projects/spring-framework/blob/master/spring-webflux/src/main/java/org/springframework/web/reactive/resource/ResourceTransformer.java

Is there a way of using Spring Webflux's reactive `ResourceChainRegistration`'s `addTransformer` (or another reactive component) to achieve the non-blocking equivalent of the following (inserting X-Ray to the start of the filter stack)? I was thinking along the lines of spring-web's `FilterRegistrationBean` being roughly similar to spring-webflux's `ResourceChainRegistration`, and spring-web's `registration` probably equates to spring-webflux's `addTransformer`?

```
  @Bean
  public FilterRegistrationBean<AWSXRayServletFilter> xrayFilter() {
    final FilterRegistrationBean<AWSXRayServletFilter> registration =
        new FilterRegistrationBean<AWSXRayServletFilter>(new AWSXRayServletFilter(tracingName));
    registration.setOrder(Ordered.HIGHEST_PRECEDENCE);
    return registration;
  }
```

I'm sure I'm at least roughly asking a question with reasonable guesses at what would be required, but I'm actually mainly wondering if this has already been done or is in-progress somewhere?

**Comment by chanchiem** [comment]
@stevegoossens 
There currently isn't any work in progress for support with the `ResourceTransformer` object. After some investigation, it looks like it would be possible to write a custom `XRayResourceTransformer` that would function similarly to our `AWSXRayServletFilter`, but perhaps not as long since there aren't as many methods to override. 

In the resource transformer interface, it looks like the `ServerWebExchange` object exposed in the `transform` method contains HTTP information about the request and response, which is basically what is needed to generate an informative `Segment`. As long as there is a distinguishable difference between when the incoming request occurs and when the outgoing response is processed, I believe that the `XRayResourceTransformer` would look fairly similar to the `AWSXRayServletFilter`

We gladly accept pull requests and would love to help guide writing this enhancement.

**Comment by lqc** [comment]
@chanchiem I'm afraid this is not that simple. I took a shot at making by own ``WebFilter`` that wraps all requests in X-Ray segment:

https://gist.github.com/lqc/4914f0f5033894c60ec28d08215edfc2

and there is a major problem here, i.e. the ``AWSXRayRecorder`` keeps global state based on ``ThreadLocal``. In reactive APIs threads can be reused mid-request. An equivalent of ``ThreadLocal`` would be something like [``Flux.context()``](https://projectreactor.io/docs/core/release/api/reactor/util/context/Context.html), but it required the API accesing it to be reactive (so ``SegmentContext`` won't fit)

Another issue would be that the default ``UDPEmitter`` uses blocking I/O, but this relatively easy to fix and impact is small as sending UDP packets is already pretty non-blocking.

**Comment by chanchiem** [comment]
The recommendation that I have is to implement a custom `SegmentContext` that utilizes the context provided by the webflux. `SegmentContext` is an interface that by defaults utilizes the thread local storage; I don't see why the global recorder can't be configured to use the custom context as part of using webflux.

**Comment by stevegoossens** [comment]
@willarmiros, is this an upcoming enhancement?

**Comment by willarmiros** [comment]
No, I've reopened the issue since it shouldn't have been closed but it is not currently on our roadmap. It's open as a feature request.

**Comment by mikeyan123** [comment]
Any progress?

**Comment by stevegoossens** [comment]
As a workaround, I've had some initial success by using Spring Sleuth with its opentracing/zipkin config, and attached x-ray to that via a zipkin extension. I could see the tracing data in aws console, but some fields like uri path were metadata instead of in the main field, so it needs a config tweak somewhere.

Point is, I could get traces with webflux into x-ray, so it's technically already possible. So that demonstrates two things:

- you could use Sleuth and some custom config to get x-ray traces
- aws-xray-sdk-java could use the same approach as Sleuth to collect trace segments in webflux

**Comment by willarmiros** [comment]
Thanks for the recommendation @stevegoossens! We're still not able to prioritize first-class WebFlux support on the SDK, however WebFlux is supported using the [OpenTelemetry Java Instrumentation](https://github.com/open-telemetry/opentelemetry-java-instrumentation#supported-java-libraries-and-frameworks). As a disclaimer, the OpenTelemetry Java Agent and X-Ray's integration with OpenTelemetry are in beta and preview states, respectively. For documentation on using the OpenTelemetry Java Agent with X-Ray as your back end, see [these docs](https://aws-otel.github.io/docs/getting-started/java-auto-instrumentation-agent).
